cardano-explorer/src/Explorer/Web/ClientTypes.hs

```diff
@@ -0,0 +1,264 @@
+
+module Explorer.Web.ClientTypes
+       ( CHash (..)
+       , CAddress (..)
+       , CTxId (..)
+       , CBlockEntry (..)
+       , CTxEntry (..)
+       , CBlockSummary (..)
+       , CAddressType (..)
+       , CAddressSummary (..)
+       , CTxBrief (..)
+       , CUtxo  (..)
+       , CNetworkAddress (..)
+       , CTxSummary (..)
+       , CBlockRange (..)
+       , CGenesisSummary (..)
+       , CGenesisAddressInfo (..)
+       , CAddressesFilter (..)
+       , CCoin
+       , CAda (..)
+       , CByteString (..)
+       , toCHash
+       ) where
+
+import qualified Prelude
+import           Universum
+
+import           Control.Monad.Error.Class (throwError)
+import           Data.Default (Default (..))
+import           Data.Fixed (Micro, showFixed)
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import           Data.Text (Text)
+import           GHC.Generics (Generic)
+import           Data.Time.Clock.POSIX (POSIXTime)
+import           Formatting (build, sformat, (%), Buildable)
+import           Servant.API (FromHttpApiData (..))
+import qualified Data.ByteString.Char8 as SB8
+import qualified Data.ByteString.Base16 as B16
+
+import Cardano.Crypto.Hash.Class
+
+-------------------------------------------------------------------------------------
+-- Hash types
+-------------------------------------------------------------------------------------
+
+-- See this page for more explanation - https://cardanodocs.com/cardano/addresses/
+-- We have the general type @AbstractHash@ for all hashes we use. It's being parametrized
+-- by two types - AbstractHash algo a - the hashing algorithm and the phantom type for
+-- extra safety (can be a @Tx@, an @Address@ and so on, ...).
+--
+-- The following types explain the situation better:
+--
+-- type AddressHash   = AbstractHash Blake2b_224
+-- type Hash          = AbstractHash Blake2b_256
+--
+-- type TxId          = Hash Tx               = AbstractHash Blake2b_256 Tx
+-- type StakeholderId = AddressHash PublicKey = AbstractHash Blake2b_224 PublicKey
+--
+-- From there on we have the client types that we use to represent the actual hashes.
+-- The client types are really the hash bytes converted to Base16 address.
+
+-- | Client hash
+newtype CHash = CHash Text
+  deriving (Show, Eq, Generic, Buildable, Hashable, NFData)
+
+-- | Client address. The address may be from either Cardano or RSCoin.
+newtype CAddress = CAddress Text
+    deriving (Show, Eq, Generic, Buildable, Hashable, NFData)
+
+-- | Client transaction id
+newtype CTxId = CTxId CHash
+    deriving (Show, Eq, Generic, Buildable, Hashable, NFData)
+
+-------------------------------------------------------------------------------------
+-- Composite types
+-------------------------------------------------------------------------------------
+
+newtype CCoin = CCoin
+    { getCoin :: Text
+    } deriving (Show, Generic, Eq)
+
+--instance NFData CCoin
+
+
+newtype CAda = CAda
+    { getAda :: Micro
```

Comment:
`Micro` is a type from the Serokell code isn't it? The DB stores ADA values denominated in `Lovelace` as a `Word64`. You should probably use `Word64` here.

Reply:
ON second thoughts, maybe. We can talk about that. Obviously the values on the web UI should display values in `ADA` (as fixed point) rather than `Lovelace`

Reply:
Or it can use `Lovelace` which is a `Word64` wrapper, and just format in ADA.

Reply:
`import           Data.Fixed (Micro, showFixed)`